### DISCUSSION TITLE
Long stall before present when adding shadow pass

### DISCUSSION THREAD

**Discussion by DGriffin91** [attackgoat/screen-13#78] [Q&A]
I have a renderer that does `shadow pass` > `main pass` > `blit to swapchain` > `present`.
All the passes use the same graph, and the main pass depends on the depth texture output of the shadow pass.
When I add the shadow pass I end up stalled at "wait for presentation" (assuming waiting on fences) for around 1ms. It seems like during this time both the CPU and GPU are fairly idle. This is with vsync disabled. Total cpu frame time without the shadow pass is around 1.8ms vs 3ms with it (when measuring without nsight/tracy).

If I remove the shadow pass, the stall seems to completely go away.

Any thoughts on what to look at or try to see what could be causing the stall?

Here's tracy:
![image](https://github.com/user-attachments/assets/41da1e42-517f-436d-b49e-d1943efd6cb1)
And nsight
![image](https://github.com/user-attachments/assets/b7125f5d-d6e4-48a8-b823-309f4b357e57)

nsight just shows this barrier
![image](https://github.com/user-attachments/assets/28aaf09a-f109-4643-8576-d40cdb607f3d)
and then the next thing is present
![image](https://github.com/user-attachments/assets/8db1ef73-a57e-475a-b394-96d2ac330b6b)

**Comment by attackgoat** [comment]
There are some improvements to swapchain synchronization in #75 (see [src/driver/swapchain.rs](https://github.com/attackgoat/screen-13/pull/75/files#diff-62f328f55abc9acb7db45f9b1b40344a87d401ba0d14fa4dd7915d34b8758197)). That branch changes `winit` app handling though and might take a tiny bit of effort to switch to (use the new `screen_13_window::Window` in place of `screen_13::EventLoop`).

Try adding lots of `profiling::scope!("blah")` first and also check how many swapchain images [the event loop was created with](https://docs.rs/screen-13/latest/screen_13/prelude/struct.EventLoopBuilder.html#method.desired_swapchain_image_count):

```
 INFO  screen_13::driver::swapchain                > Swapchain 1024x1024 B8G8R8A8_UNORM FIFO_RELAXEDx3
```

You've probably already found those though. I'll see if I can learn anything from these screenshots over the next few days.

**Reply by DGriffin91** [reply]
Tried the Ash 0.38 branch. Switching over was easy as I'm not using the built-in event loop. It appears the stall issue still occurs. Everything else seems to be working on ash 0.38 though! Nice work!

Realized it was using mailbox. Also tried immediate with the same issue. The lowest I can get is `Swapchain 1920x1080 B8G8R8A8_UNORM IMMEDIATEx2.` (but still have the ~1ms stall)

According to this `warn!` I'm often spending over 1ms waiting for fences in the ash `device.wait_for_fences` call: https://github.com/attackgoat/screen-13/blob/c5f9e81eaf47bf0619dc98d5dbdb3daaca73eb04/src/driver/device.rs#L402

I've also included a flame graph from vtune showing this wait.

![image](https://github.com/user-attachments/assets/36814ae9-cb1e-408a-8382-9dbc6a299940)
![image](https://github.com/user-attachments/assets/24398eed-6545-4e5c-9cde-b7b7db17f496)

**Reply by attackgoat** [reply]
If your project is not using the built-in `EventLoop` (now `Window`) abstractions then it certainly is using `Display::new` to create the present-helper type. What value is passed for `cmd_buf_count`?

It looks like the display helper is pulling a command buffer out of the pile it keeps around, but the pulled buffer has not finished submission yet.

`cmd_buf_count`:

> Specifies the number of in-flight command buffers, which should be greater
than or equal to the desired swapchain image count.
>
> More command buffers mean less time waiting for previously submitted frames to complete, but
more memory in use.
>
> Generally a value of one or two greater than desired image count produces the smoothest
animation.

**Reply by DGriffin91** [reply]
Yep, I'm using Display::new. cmd_buf_count is currently 5. It looks like with setting it to 2 or 12 it still stalls. It's also worth noting that it looks like it doesn't stall every frame, just most of them. The frames where it doesn't stall will both not have the gpu idle time while being shorter overall.

**Reply by DGriffin91** [reply]
I have a hunch that part of what is happening is that the waiting on the fence is the CPU waiting for the GPU to finish, and the stall on the GPU is the GPU then waiting for the CPU to give it more work. I think there may be something else also going on too. But I think a big portion of the time might be that.

**Reply by attackgoat** [reply]
The stall during `Display::wait_for_fence` is a certain indication that Vulkan is stalling due to the resource barriers generated by a previous frame. That fence guards against the previous submission to the round-robbin selected command buffer having not finished yet.

Setting `cmd_buf_count` to `12` means there should be enough command buffers sitting around that the selected one should be idle.

Setting the swapchain to immediate means that Vulkan won't wait for presentation, but it will still wait for rendering commands. If you are able to increase the swapchain count that could show you if it is waiting on the image to be rendered.

The problem is likely a single resource used in every frame - although two resources used as ping-pong buffers could potentially cause the same issue. Anything leased from a pool is not the issue, it's almost certainly something like an `Image::create(..)`/`Buffer::create(..)` resource. If you're using the `AliasPool` wrapper it could cause it - that allows you to lease the same resource in multiple places without having to keep track of resources like manually created ones.

Unrelated, but the `libloading`/`hassle_rs` and large amount of work being done in the `record_subpass(..)` (or other record callback) seem kind of odd - does vtune show aggregate frames or a single whole frame?

**Reply by DGriffin91** [reply]
Tried a swapchain count of 8, in tracy it looks about the same as 3 and the overall time is about the same too.

I was using Image::create but only for loading the initial assets. All the other buffers and images appear to be using the pool, and I just setup the initial image asset loading to also use the pool. Overall frame rate seems the same, but startup time is maybe faster now. I'm not using `AliasPool` currently.

vtune isn't aware of frames and just looks at the whole execution. The libloading/hassle_rs stuff is just during startup. I'm not sure what specific thing you are referring to with record_subpass, is it the record_node_passes? It does seem like that reorder step take a long time.

**Reply by attackgoat** [reply]
> It does seem like that reorder step take a long time.

I was mistaken; I saw all the iterator stuff and thought it was coming from one of those callbacks but it appears to be the iterator functions called by reorder. I need to make another pass at making the reorder step more efficient or allowing it to be skipped (you could comment out the line which calls it, but it is unrelated).

I would try running your program on multiple graphics cards from different vendors and see if you get the same behavior. I'm more than happy to help debug your code if you could provide it; I keep a small collection of machines and cards for this purpose. My email is john@attackgoat.com and I sign NDAs and can do this at no cost to you. Whatever I learn might improve _Screen 13_.

However, I also wouldn't try to fix 1ms unless it was visible in builds during the polish portion of your project. This might disappear as you build out the rendering portion of your project and stuff more into the GPU pipeline.

**Reply by DGriffin91** [reply]
Thanks for the offer! That would be awesome! I'll send over an email when I have the opportunity.

**Reply by DGriffin91** [reply]
@attackgoat Just a heads up. I'm taking brief break from this to participate in https://itch.io/jam/bevy-jam-5. I'll be back working on this next week!

**Reply by DGriffin91** [reply]
So it looks like the wdm.exe stalls go away if I use full screen instead of windowed. In both cases things are looking a lot better now. I'm not sure exactly what changed, I've updated lots of things, but if I go back to an old commit I still don't see the same huge stall at the end, but a much smaller one.

Fullscreen:
![fullscreen](https://github.com/user-attachments/assets/9973f150-9035-4a1c-960d-318d62df43c2)

Windowed:
![windowed](https://github.com/user-attachments/assets/540a35e9-9e06-4dd0-b52c-cf33815cfd49)
